wandb/sdk/wandb_run.py

```diff
@@ -1959,15 +1977,153 @@ def _on_ready(self) -> None:
+            "source_type": "repo",
+            "remote": self._remote_url,
+            "commit": self._last_commit,
+            "entrypoint": [
+                sys.executable.split("/")[-1],
+                self._settings.program_relpath,
+            ],
+            "input_types": input_types,
+            "output_types": output_types,
+            "runtime": self._settings._python,
+        }
+        with job_artifact.new_file("source_info.json") as f:
+            f.write(json.dumps(source_info))
+
+        artifact = self.log_artifact(job_artifact)
+        return artifact
+
+    def _create_artifact_job(
+        self,
+        input_types: Dict[str, Any],
+        output_types: Dict[str, Any],
+        installed_packages_list: List[str],
+    ) -> "Artifact":
+        assert self._code_artifact is not None
+        assert self._run_obj is not None
+        self._code_artifact.wait()
+        sequence_name = self._code_artifact.name.split(":")[0]
+        tag = self._code_artifact.version
+        name = f"job-{sequence_name}"
+        job_artifact = wandb.Artifact(name, type="job")
+
+        with job_artifact.new_file("requirements.frozen.txt") as f:
+            f.write("\n".join(installed_packages_list))
+
+        source_info = {
+            "_version": "v0",
+            "source_type": "artifact",
+            "artifact": f"wandb-artifact://{self._run_obj.entity}/{self._run_obj.project}/{sequence_name}:{tag}",
+            "entrypoint": [
+                sys.executable.split("/")[-1],
+                self._settings.program,
+            ],
+            "input_types": input_types,
+            "output_types": output_types,
+            "runtime": self._settings._python,
+        }
+        with job_artifact.new_file("source_info.json") as f:
+            f.write(json.dumps(source_info))
+
+        artifact = self.log_artifact(job_artifact)
+        return artifact
+
+    def _create_container_job(
+        self, input_types: Dict[str, Any], output_types: Dict[str, Any]
+    ) -> "Artifact":
+        name = wandb.util.make_artifact_name_safe(f"job-{os.getenv('WANDB_DOCKER')}")
+        job_artifact = wandb.Artifact(name, type="job")
+
+        source_info = {
+            "_version": "v0",
+            "source_type": "image",
+            "image": os.getenv("WANDB_DOCKER"),
+            "input_types": input_types,
+            "output_types": output_types,
+            "runtime": self._settings._python,
+        }
+        with job_artifact.new_file("source_info.json") as f:
+            f.write(json.dumps(source_info))
+
+        artifact = self.log_artifact(job_artifact)
+        return artifact
+
     def _on_finish(self) -> None:
         trigger.call("on_finished")
-
         # populate final import telemetry
         with telemetry.context(run=self) as tel:
             self._telemetry_imports(tel.imports_finish)
 
         if self._run_status_checker:
             self._run_status_checker.stop()
+        if (
```

Comment:
just realized that in service it will require explicitly calling `finish` or having some logic in when we finish a stream

Reply:
do we in service have any mechanism to ensure something is callled?

Reply:
hmm... in service when we clean up the outstanding run (i.e. runs that didn't call finish explicitly) we don't have access to the run instance so we can't use the run's state...